engine/crates/engine-v2/src/plan/planner/boundary_planner.rs

```diff
@@ -0,0 +1,592 @@
+use std::{
+    borrow::Cow,
+    collections::{hash_map::Entry, HashMap, HashSet},
+};
+
+use schema::{FieldId, FieldResolverWalker, FieldSet, FieldSetItem, ResolverId, ResolverWalker};
+
+use crate::{
+    plan::{
+        attribution::{AttributionBuilder, ExtraField},
+        ChildPlan, EntityType, FlatTypeCondition, PlanId,
+    },
+    request::{
+        BoundFieldId, BoundSelectionSetId, FlatField, FlatSelectionSet, FlatSelectionSetId, FlatSelectionSetWalker,
+        GroupForFieldId, OperationWalker, QueryPath, SelectionSetType,
+    },
+    response::{ReadField, ReadSelectionSet},
+};
+
+use super::{AttributionLogic, ExpectedType, Planner, PlanningError, PlanningResult};
+
+pub(super) struct PlanBoundaryChildrenPlanner<'op, 'a> {
+    planner: &'a mut Planner<'op>,
+    walker: OperationWalker<'op>,
+    maybe_parent: Option<PlanBoundaryParent<'op, 'a, 'a>>,
+    children: Vec<ChildPlan>,
+}
+
+pub(super) struct PlanBoundaryParent<'op, 'plan, 'ctx> {
+    pub plan_id: PlanId,
+    pub path: &'ctx QueryPath,
+    pub logic: AttributionLogic<'op>,
+    pub attribution: &'plan mut AttributionBuilder,
+    pub flat_selection_set: FlatSelectionSetWalker<'op>,
+}
+
+#[derive(Debug)]
+enum BoundaryField<'op> {
+    Planned {
+        field: GroupForFieldId<'op>,
+        subselection: Option<BoundarySelectionSet<'op>>,
+    },
+    Extra {
+        plan_id: PlanId,
+        resolver_id: ResolverId,
+        field: ExtraBoundaryField,
+    },
+}
+
+#[derive(Debug)]
+struct BoundarySelectionSet<'op> {
+    id: FlatSelectionSetId,
+    fields: HashMap<FieldId, BoundaryField<'op>>,
+}
+
+#[derive(Debug, Clone)]
+pub(in crate::plan) struct ExtraBoundaryField {
+    pub extra_field: ExtraField<ExtraBoundarySelectionSet>,
+    pub read: bool,
+}
+
+#[derive(Debug, Clone)]
+pub(in crate::plan) struct ExtraBoundarySelectionSet {
+    pub ty: SelectionSetType,
+    pub fields: HashMap<FieldId, ExtraBoundaryField>,
+}
+
+impl<'op, 'a> PlanBoundaryChildrenPlanner<'op, 'a> {
+    pub fn new(planner: &'a mut Planner<'op>, maybe_parent: Option<PlanBoundaryParent<'op, 'a, 'a>>) -> Self {
+        let walker = planner.default_operation_walker();
+        PlanBoundaryChildrenPlanner {
+            planner,
+            walker,
+            maybe_parent,
+            children: vec![],
+        }
+    }
+
+    pub fn plan_children(
```

Comment:
Given that this function is like 200 lines long (and I assume it calls into the other 400 lines of this file) a docstring summarising its purpose would be useful

Reply:
Maybe even splitting it into its own module with flat functions, smaller functions and a healthy amount of rustdoc?